File: YODA/VirtualMachine.cs
// ReSharper disable InconsistentNaming
namespace SimpleInstructionMachine;

public class VirtualMachine(bool Debug)
{
    private static class Mask
    {
        public const byte Halt = 0b0000;
        public const byte SaveToFile = 0b0001;
        public const byte LoadFromFile = 0b0010;
        public const byte Write = 0b0011;
        public const byte Add = 0b0100;
        public const byte Sub = 0b0101;
        public const byte Inc = 0b0110;
        public const byte Dec = 0b0111;
        public const byte Nop = 0b1000;
        public const byte JumpIfZero = 0b1001;
        public const byte Copy = 0b1010;
    }

    private static class OpCode
    {
        //1 means immediate rather than memory (0)
        
        //Values can either be
        //      M - Memory (0) or Immediate (1)
        //      D - Indirect (0) or Direct (1) 
        
        public const byte Halt = 0b0000_0000;   // 00
        public const byte Wait = 0b0000_0001;   // 00

        
        public const byte SaveToFileMMM = 0b0001_0000;  //16
        public const byte SaveToFileMMI = 0b0001_0001;  //17
        public const byte SaveToFileMIM = 0b0001_0010;  //18
        public const byte SaveToFileMII = 0b0001_0011;  //19
        public const byte SaveToFileIMM = 0b0001_0100;  //20
        public const byte SaveToFileIMI = 0b0001_0101;  //21
        public const byte SaveToFileIIM = 0b0001_0110;  //22
        public const byte SaveToFileIII = 0b0001_0111;  //23
        
        public const byte LoadFromFileMM = 0b0010_0000;  //32
        public const byte LoadFromFileMI = 0b0010_0001;  //33
        public const byte LoadFromFileIM = 0b0010_0010;  //34
        public const byte LoadFromFileII = 0b0010_0011;  //35
        
        public const byte WriteMM = 0b0011_0000;  //48
        public const byte WriteMI = 0b0011_0001;  //49
        public const byte WriteIM = 0b0011_0010;  //50
        
        public const byte AddMMD = 0b0100_0000;  //64
        public const byte AddMMI = 0b0100_0001;  //65
        public const byte AddMID = 0b0100_0010;  //66
        public const byte AddMII = 0b0100_0011;  //67
        public const byte AddIMD = 0b0100_0100;  //68
        public const byte AddIMI = 0b0100_0101;  //69
        public const byte AddIID = 0b0100_0110;  //70
        public const byte AddIII = 0b0100_0111;  //71

        public const byte SubMMD = 0b0101_0000;  //80
        public const byte SubMMI = 0b0101_0001;  //81
        public const byte SubMID = 0b0101_0010;  //82
        public const byte SubMII = 0b0101_0011;  //83
        public const byte SubIMD = 0b0101_0100;  //84
        public const byte SubIMI = 0b0101_0101;  //85
        public const byte SubIID = 0b0101_0110;  //86
        public const byte SubIII = 0b0101_0111;  //87
        
        public const byte IncM = 0b0110_0000;  //96
        public const byte IncI = 0b0110_0001;  //97
        
        public const byte DecM = 0b0111_0000;  //112
        public const byte DecI = 0b0111_0001;  //113
        
        public const byte Nop = 0b1000_0000; //128

        public const byte JumpIfZeroMM = 0b1001_0000;  //144
        public const byte JumpIfZeroMI = 0b1001_0001;  //145
        public const byte JumpIfZeroIM = 0b1001_0010;  //146
        public const byte JumpIfZeroII = 0b1001_0011;  //147
        
        public const byte CopyMM = 0b1010_0000;  //160
        public const byte CopyMI = 0b1010_0001;  //161
        public const byte CopyIM = 0b1010_0010;  //162
        public const byte CopyII = 0b1010_0011;  //163
    }
    
    private const byte LCD_0 = 0xF8;
    private const byte LCD_1 = 0xF9;
    private const byte LCD_2 = 0xFA;
    private const byte LCD_3 = 0xFB;
    private const byte LCD_4 = 0xFC;

    private const byte ControlFlags = 0xFD;
        
    
    private readonly byte[] _memory = new byte[1 + byte.MaxValue];

    private int _instructionPointer = 0;

    private string _folder = ".";
    
    public async Task Run(string folderPath)
    {
        
        _folder = folderPath;
        
        await Initialise();

        var halted = false;
        while (!halted)
        {
            var opCode = _memory[_instructionPointer];
            try
            {
                switch (opCode >> 4)
                {
                    case 0:
                        switch (opCode)
                        {
                            case OpCode.Halt:
                            {
                                halted = true;
                                continue;
                            }
                            case OpCode.Wait:
                            {
                                await Wait();
                                continue;
                            }
                            default:
                                throw new Exception("Unknown command " + opCode);
                        }

                    case Mask.SaveToFile:
                        await SaveToFile(opCode);
                        continue;
                    case Mask.LoadFromFile:
                        await LoadFromFile(opCode);
                        continue;
                    case Mask.Write:
                        Write(opCode);
                        continue;
                    case Mask.Add:
                        Add(opCode);
                        continue;
                    case Mask.Sub:
                        throw new Exception("Due to lack of time this method has not been implemented");
                    case Mask.Inc:
                        Inc(opCode);
                        continue;
                    case Mask.Dec:
                        Dec(opCode);
                        continue;
                    case Mask.Nop:
                        Nop();
                        break;
                    case Mask.JumpIfZero:
                        JumpIfZero(opCode);
                        break;
                    case Mask.Copy:
                        Copy(opCode);
                        break;
                    default:
                        throw new Exception("Unknown command " + opCode);
                }
                
                Console.WriteLine("\n\nProgram completed successfully");
                
            }
            catch (Exception e)
            {
                await Console.Error.WriteLineAsync($"Your program has crashed,  things aren't looking to good for the space craft.\n");
                await Console.Error.WriteLineAsync($"{e.Message}");
                await Console.Error.WriteLineAsync($"Instruction Pointer: {_instructionPointer}");
                await Console.Error.WriteLineAsync($"Opcode: {opCode}\n");
                
                // Dump as bytes
                await File.WriteAllBytesAsync("crash_dump", _memory);

                // Dump as text
                await using var textFile = File.CreateText("crash_dump.txt");
                for (var i = 0; i < _memory.Length; i++)
                {
                    if (i == _instructionPointer)
                        await textFile.WriteLineAsync($"{i:X2}   {_memory[i]}    <---- INSTRUCTION POINTER");
                    else
                        await textFile.WriteLineAsync($"{i:X2}   {_memory[i]}");
                }

                await textFile.FlushAsync();                    
                
                await Console.Error.WriteLineAsync($"A crash dump containing all the memory has been written to : crash_dump and crash_dump.txt");
                return;
            }
        }
    }

    private async Task Initialise()
    {
        Array.Fill(_memory,(byte)0);
        _instructionPointer = 0;

        var filename = Path.Combine(_folder, "boot");
        if (File.Exists(filename))
        {
            var fileContents = await File.ReadAllBytesAsync(filename);
            if (fileContents.Length > _memory.Length)
                throw new Exception($"The boot file is too large. It is {fileContents.Length} bytes long,  which exceeds the maximum allowed of {_memory.Length} bytes");

            fileContents.CopyTo(_memory, 0);
            
            Console.WriteLine($"\nMemory has been initialised using the boot file ({filename}).");
        }
        else
        {
            Console.WriteLine("\nNo boot file found.");
        }
    }

    private string FilenameFromFileNumber(byte fileNumber)
    {
        return fileNumber switch
        {
            < 8 => Path.Combine(_folder, $"{fileNumber}"),
            < 16 => Path.Combine(_folder, $"{fileNumber}.txt"),
            _ => throw new Exception($"Unknown file {fileNumber}.  Binary files are between 0 and 7.   Text files are between 8 and 15")
        };
    }
    
    /// <summary>
    /// SaveToFile FileNumber SourceLocation Length
    /// </summary>
    private async Task SaveToFile(int opCode)
    {
        var fileNumber = Read(_instructionPointer + 1, opCode, 2);
        var sourceLocation = Read(_instructionPointer + 2 , opCode, 1);
        var length = Read(_instructionPointer + 3 , opCode, 0);
        
        await File.WriteAllBytesAsync(FilenameFromFileNumber(fileNumber), _memory[sourceLocation..(sourceLocation+length)]);
        
        _instructionPointer += 4;
    }
    
    
    /// <summary>
    /// LoadFromFile FileNumber SourceLocation Length
    /// </summary>
    private async Task LoadFromFile(int opCode)
    {
        var fileNumber = Read(_instructionPointer + 1, opCode, 1);
        var targetLocation = Read(_instructionPointer + 2 , opCode, 0);
        
        var fileContents = await File.ReadAllBytesAsync(FilenameFromFileNumber(fileNumber));
        if (fileContents.Length + targetLocation > _memory.Length)
            throw new Exception("File too large");
        fileContents.CopyTo(_memory, targetLocation );
        
        _instructionPointer += 3;
    }
    
    
    
    /// <summary>
    /// Write [Location] Value
    /// </summary>
    private void Write(int opCode)
    {
        var location = Read(_instructionPointer + 1, opCode, 1);
        var value = Read(_instructionPointer + 2 , opCode, 0);

        if (Debug) Console.WriteLine($"Write {value} into {location:X2}");
        
        if (location == ControlFlags)
        {
           var diffs = _memory[location] ^ value;
           if ((diffs & 1) == 1)
           {
               //bit 0 has changed, refresh the LCD display
               Console.Write(ToChar(_memory[LCD_0]));
               Console.Write(ToChar(_memory[LCD_1]));
               Console.Write(ToChar(_memory[LCD_2]));
               Console.Write(ToChar(_memory[LCD_3]));
               Console.WriteLine(ToChar(_memory[LCD_4]));

               char ToChar(byte b)
               {
                   if (b == 0x00)
                       return ' ';
                   else
                       return (char)b;
               }

           }
        }
        
        _memory[location] = value;
        _instructionPointer += 3;
    }
    
    /// <summary>
    /// Add LHS RHS Total
    /// </summary>
    private void Add(int opCode)
    {
        var lhs = Read(_instructionPointer + 1, opCode, 2);
        var rhs = Read(_instructionPointer + 2 , opCode, 1);
        var location = Read(_instructionPointer + 3 , opCode, 0);
        
        _memory[location] = (byte)(lhs+rhs); // Can overflow
        _instructionPointer += 4;
    }
    
    
    /// <summary>
    /// Inc [Location]
    /// </summary>
    private void Inc(int opCode)
    {
        var location = Read(_instructionPointer + 1 , opCode, 0);
        _memory[location]++;
        _instructionPointer += 2;
    }
    
    /// <summary>
    /// Dec [Location]
    /// </summary>
    private void Dec(int opCode)
    {
        var location = Read(_instructionPointer + 1 , opCode, 0);
        
        if (Debug) Console.WriteLine($"DEC: Decreasing value in {location:x8} from {_memory[location]} to {(_memory[location])-1}");
        
        _memory[location]--;
        _instructionPointer += 2;
    }
    
    /// <summary>
    /// Nop
    /// </summary>
    private void Nop()
    {
        _instructionPointer++;
    }

    private async Task Wait()
    {
        if (Debug) Console.WriteLine("Wait");
        await Task.Delay(100);
        _instructionPointer++;
    }

    /// <summary>
    /// JumoIfZero ValueToCheck Address
    /// </summary>
    private void JumpIfZero(int opCode)
    {
        var valueToCheck = Read(_instructionPointer + 1 , opCode, 1);
        var locationToJumpTo = Read(_instructionPointer + 2 , opCode, 0);

        if (Debug) Console.WriteLine($"JumpIfZero ({opCode:b8}) - jump to {locationToJumpTo:X2} if {valueToCheck} is 0");
         
        if (valueToCheck == 0)
            _instructionPointer = locationToJumpTo;
        else
            _instructionPointer += 3;
    }

    /// <summary>
    /// Copy [SourceLocation] [TargetLocation]
    /// </summary>
    private void Copy(int opCode)
    {
        var value = Read(_instructionPointer + 1 , opCode, 1);
        var targetAddress = Read(_instructionPointer + 2 , opCode, 0);

        _memory[targetAddress] = value;
        
        _instructionPointer += 3;
    }

    private byte Read(int location,int opCode, int mask)
    {
        //isSet means immediate rather than memory
        var isSet = (((opCode & 0b0000_1111) >> mask) & 1) == 1;
        // Console.WriteLine($"{opCode:b8} {(opCode & 0b0000_1111):b8} mask {mask} {(((opCode & 0b0000_1111) >> mask) & 1 ):b8} {isSet} ");
        
        if (location >= _memory.Length)
            throw new Exception("Illegal memory location " + location);

        if (isSet)
        {
            // Console.WriteLine($"{opCode:b8}, {mask}, {_memory[location]:x8}");  
            return _memory[location];
        }

        var reference = _memory[location];
        if (reference >= _memory.Length)
            throw new Exception("Illegal de-referenced memory location " + location);
        
        // Console.WriteLine($"{opCode:b8}, {mask}, {reference:x8}, {_memory[reference]}");

        
        return  _memory[reference];
    } 
}

File: YODA_TypeScript/VirtualMachine.ts
import * as fs from "fs/promises";
import * as path from "path";
import { KnownMemory } from "./KnownMemory";
import { Mask } from "./Mask";
import { OpCode } from "./OpCodes";


// === VIRTUAL MACHINE ===
export class VirtualMachine {
  private readonly memory: Uint8Array = new Uint8Array(1 + 0xff);
  private instructionPointer: number = KnownMemory.APP_DATA_BOTTOM;
  private stackHeadPointer: number = KnownMemory.STACK_BOTTOM;
  private interruptsEnabled: boolean = false;
  private folder: string = ".";
  private debug: boolean;

  // Interrupt state
  private lastKey: string | null = null;

  constructor(debug: boolean) {
    this.debug = debug;

    // Setup stdin for keypress interrupts
    if (process.stdin.isTTY) {
      process.stdin.setRawMode(true);
      process.stdin.resume();
      process.stdin.on("data", (chunk: Buffer) => {
        this.lastKey = chunk.toString("utf8");
        if (this.lastKey === "\u0003") { // ^C
          console.log("\nCTRL+C detected — shutting down VM.");
          process.exit(0);
        }
      });
    }
  }

  async run(folderPath: string): Promise<void> {
    this.folder = folderPath;
    await this.boot();

    let halted = false;
    while (!halted) {
      // 🔹 Interrupt handling
      if (this.interruptsEnabled && this.lastKey) {
        const key = this.lastKey;
        this.lastKey = null; // consume key

        if (key === "\u001b[D") { // Left arrow = ESC [ D
          this.pushToStack(this.instructionPointer);
          this.instructionPointer = this.memory[KnownMemory.IVT_LEFT_ARROW]!;
        } else if (key === "\u001b[C") { // Right arrow = ESC [ C
          this.pushToStack(this.instructionPointer);
          this.instructionPointer = this.memory[KnownMemory.IVT_RIGHT_ARROW]!;
        }
      }


      const opCode = this.memory[this.instructionPointer]!;
      try {
        switch (opCode >> 4) {
          case Mask.Misc:
            switch (opCode) {
              case OpCode.Halt: halted = true; continue;
              case OpCode.Wait: await this.wait(); continue;
              case OpCode.Nop: this.nop(); continue;
              case OpCode.Sif: this.sif(); continue;
              case OpCode.Cif: this.cif(); continue;
              case OpCode.Ret: this.ret(); continue;
              default: throw new Error("Unknown command " + opCode);
            }
          case Mask.SaveToFile: await this.saveToFile(opCode); continue;
          case Mask.LoadFromFile: await this.loadFromFile(opCode); continue;
          case Mask.Write: this.write(opCode); continue;
          case Mask.Add: this.add(opCode); continue;
          case Mask.Sub: throw new Error("Sub not implemented");
          case Mask.Inc: this.inc(opCode); continue;
          case Mask.Dec: this.dec(opCode); continue;
          case Mask.JumpIfZero: this.jumpIfZero(opCode); continue;
          case Mask.JumpWithReturn: this.jumpWithReturn(opCode); continue;
          default: throw new Error("Unknown command " + opCode);
        }
      } catch (e: any) {
        console.error("Your program has crashed!");
        console.error(e.message);
        console.error(`Instruction Pointer: ${this.instructionPointer}`);
        console.error(`Opcode: ${opCode}`);

        await fs.writeFile("crash_dump", this.memory);

        const lines: string[] = [];
        for (let i = 0; i < this.memory.length; i++) {
          if (i === this.instructionPointer)
            lines.push(`${i.toString(16).padStart(2, "0")}   ${this.memory[i]}    <---- INSTRUCTION POINTER`);
          else
            lines.push(`${i.toString(16).padStart(2, "0")}   ${this.memory[i]}`);
        }
        await fs.writeFile("crash_dump.txt", lines.join("\n"));
        return;
      }
    }

    console.log("\n\nProgram completed successfully");
    process.exit(0);
  }

  // === STACK ===
  private pushToStack(value: number) {
    this.memory[this.stackHeadPointer--] = value;
  }
  private popFromStack(): number {
    this.stackHeadPointer++;
    if (this.stackHeadPointer > KnownMemory.STACK_BOTTOM) throw new Error("Stack underflow");
    return this.memory[this.stackHeadPointer]!;
  }

  // === BOOT ===
  private async boot() {
    this.memory.fill(0);
    this.instructionPointer = KnownMemory.APP_DATA_BOTTOM;
    this.interruptsEnabled = false;
    this.stackHeadPointer = KnownMemory.STACK_BOTTOM;

    const filename = path.join(this.folder, "boot");
    try {
      const fileContents = await fs.readFile(filename);
      if (fileContents.length > this.memory.length)
        throw new Error(`Boot file too large: ${fileContents.length} > ${this.memory.length}`);
      fileContents.copy(this.memory, 0);
      console.log(`\nMemory initialised using boot file (${filename}).`);
    } catch {
      console.log("\nNo boot file found.");
    }
  }

  // === FILE IO ===
  private filenameFromFileNumber(fileNumber: number): string {
    if (fileNumber < 8) return path.join(this.folder, `${fileNumber}`);
    if (fileNumber < 16) return path.join(this.folder, `${fileNumber}.txt`);
    throw new Error(`Unknown file ${fileNumber}`);
  }

  private async saveToFile(opCode: number) {
    const fileNumber = this.read(this.instructionPointer + 1, opCode, 2);
    const sourceLocation = this.read(this.instructionPointer + 2, opCode, 1);
    const length = this.read(this.instructionPointer + 3, opCode, 0);

    if (this.debug)
      console.log(`${this.instructionPointer.toString(16)} SaveToFile:: Writing ${length} bytes from ${sourceLocation.toString(16)}.`);

    await fs.writeFile(
      this.filenameFromFileNumber(fileNumber),
      this.memory.slice(sourceLocation, sourceLocation + length)
    );

    this.instructionPointer += 4;
  }

  private async loadFromFile(opCode: number) {
    const fileNumber = this.read(this.instructionPointer + 1, opCode, 1);
    const targetLocation = this.read(this.instructionPointer + 2, opCode, 0);

    const fileContents = await fs.readFile(this.filenameFromFileNumber(fileNumber));
    if (fileContents.length + targetLocation > this.memory.length)
      throw new Error("File too large");
    fileContents.copy(this.memory, targetLocation);

    if (this.debug)
      console.log(`${this.instructionPointer.toString(16)} LoadFromFile:: Reading into ${targetLocation.toString(16)}.`);

    this.instructionPointer += 3;
  }

  // === MEMORY OPS ===
  private write(opCode: number) {
    const location = this.read(this.instructionPointer + 1, opCode, 1);
    const value = this.read(this.instructionPointer + 2, opCode, 0);

    if (this.debug)
      console.log(`${this.instructionPointer.toString(16)} Write:: ${value} into ${location.toString(16)}`);

    this.updateScreenIfRequired(location, value);
    this.memory[location] = value;
    this.instructionPointer += 3;
  }

  private updateScreenIfRequired(location: number, value: number) {
    if (location === KnownMemory.ControlFlags) {
      if ((this.memory[location]! & 1) === 0 && (value & 1) === 1) {
        console.log("---------------------");
        console.log(
          `| ${this.toChar(this.memory[KnownMemory.LCD_0]!)} | ${this.toChar(this.memory[KnownMemory.LCD_1]!)} | ${this.toChar(this.memory[KnownMemory.LCD_2]!)} | ${this.toChar(this.memory[KnownMemory.LCD_3]!)} | ${this.toChar(this.memory[KnownMemory.LCD_4]!)} |`
        );
        console.log("---------------------");
      }
    }
  }

  private toChar(b: number): string {
    return b === 0x00 ? " " : String.fromCharCode(b);
  }

  private add(opCode: number) {
    const lhs = this.read(this.instructionPointer + 1, opCode, 2);
    const rhs = this.read(this.instructionPointer + 2, opCode, 1);
    const location = this.read(this.instructionPointer + 3, opCode, 0);

    if (this.debug)
      console.log(`${this.instructionPointer.toString(16)} Add:: ${lhs}+${rhs} => ${location.toString(16)}`);
    this.memory[location] = (lhs + rhs) & 0xff;
    this.instructionPointer += 4;
  }

  private inc(opCode: number) {
    const location = this.read(this.instructionPointer + 1, opCode, 0);
    if (this.debug)
      console.log(`${this.instructionPointer.toString(16)} Inc:: ${this.memory[location]} -> ${this.memory[location]! + 1}`);
    this.memory[location]!++;
    this.instructionPointer += 2;
  }

  private dec(opCode: number) {
    const location = this.read(this.instructionPointer + 1, opCode, 0);
    if (this.debug)
      console.log(`${this.instructionPointer.toString(16)} Dec:: ${this.memory[location]} -> ${this.memory[location]! - 1}`);
    this.memory[location]!--;
    this.instructionPointer += 2;
  }

  // === FLOW CONTROL ===
  private nop() {
    if (this.debug) console.log(`${this.instructionPointer.toString(16)} Nop::`);
    this.instructionPointer++;
  }

  private sif() {
    if (this.debug) console.log(`${this.instructionPointer.toString(16)} Sif:: Was ${this.interruptsEnabled}`);
    this.interruptsEnabled = true;
    this.instructionPointer++;
  }

  private cif() {
    if (this.debug) console.log(`${this.instructionPointer.toString(16)} Cif:: Was ${this.interruptsEnabled}`);
    this.interruptsEnabled = false;
    this.instructionPointer++;
  }

  private async wait() {
    if (this.debug) console.log(`${this.instructionPointer.toString(16)} Wait::`);
    await new Promise((r) => setTimeout(r, 100));
    this.instructionPointer++;
  }

  private ret() {
    const gotoAddress = this.popFromStack();
    if (this.debug)
      console.log(`${this.instructionPointer.toString(16)} Ret:: to ${gotoAddress.toString(16)}`);
    this.instructionPointer = gotoAddress;
  }

  private jumpIfZero(opCode: number) {
    const valueToCheck = this.read(this.instructionPointer + 1, opCode, 1);
    const locationToJumpTo = this.read(this.instructionPointer + 2, opCode, 0);
    if (this.debug)
      console.log(`${this.instructionPointer.toString(16)} JumpIfZero:: to ${locationToJumpTo.toString(16)} if ${valueToCheck} == 0`);
    if (valueToCheck === 0) this.instructionPointer = locationToJumpTo;
    else this.instructionPointer += 3;
  }

  private jumpWithReturn(opCode: number) {
    const locationToJumpTo = this.read(this.instructionPointer + 1, opCode, 0);
    if (this.debug)
      console.log(`${this.instructionPointer.toString(16)} JumpWithReturn:: to ${locationToJumpTo.toString(16)}`);
    this.pushToStack(this.instructionPointer + 2);
    this.instructionPointer = locationToJumpTo;
  }

  // === MEMORY READ ===
  private read(location: number, opCode: number, mask: number): number {
    const isSet = (((opCode & 0x0f) >> mask) & 1) === 1;
    if (location >= this.memory.length) throw new Error("Illegal memory location " + location);

    if (isSet) return this.memory[location]!;

    const reference = this.memory[location]!;
    if (reference >= this.memory.length) throw new Error("Illegal deref location " + location);

    return this.memory[reference]!;
  }
}
